k-distribution/include/ocaml/hooks.ml

```diff
@@ -0,0 +1,717 @@
+    | _ -> raise Not_implemented
+  let hook_read c _ _ _ _ = match c with
+      [Int fd], [Int len] -> unix_error (fun () -> let l = (Z.to_int len) in
+        let b = Bytes.create l in let _ = Unix.read (Hashtbl.find file_descriptors fd) b 0 l in [String (Bytes.to_string b)])
+    | _ -> raise Not_implemented
+  let hook_seek c _ _ _ _ = match c with
+      [Int fd], [Int off] -> unix_error (fun () -> let o = (Z.to_int off) in let _ = Unix.lseek (Hashtbl.find file_descriptors fd) o Unix.SEEK_SET in [])
+    | _ -> raise Not_implemented
+  let hook_seekEnd c _ _ _ _ = match c with
+      [Int fd], [Int off] -> unix_error (fun () -> let o = (Z.to_int off) in let _ = Unix.lseek (Hashtbl.find file_descriptors fd) o Unix.SEEK_END in [])
+    | _ -> raise Not_implemented
+  let hook_tell c _ _ _ _ = match c with
+      [Int fd] -> unix_error (fun () -> [Int (Z.of_int (Unix.lseek (Hashtbl.find file_descriptors fd) 0 Unix.SEEK_CUR))])
+    | _ -> raise Not_implemented
+  let hook_write c _ _ _ _ = match c with
+      [Int fd], [String s] -> unix_error (fun () -> let b = Bytes.of_string s in let _ = Unix.write (Hashtbl.find file_descriptors fd) b 0 (Bytes.length b) in [])
+    | _ -> raise Not_implemented
+  let hook_lock c _ _ _ _ = match c with
+      [Int fd], [Int len] -> unix_error (fun () -> Unix.lockf (Hashtbl.find file_descriptors fd) Unix.F_LOCK (Z.to_int len); [])
+    | _ -> raise Not_implemented
+  let hook_unlock c _ _ _ _ = match c with
+      [Int fd], [Int len] -> unix_error (fun () -> Unix.lockf (Hashtbl.find file_descriptors fd) Unix.F_ULOCK (Z.to_int len); [])
+    | _ -> raise Not_implemented
+
+  let log_files = Hashtbl.create 2
+
+  let hook_log c _ _ _ _ = match c with
+      [String path], [String txt] ->
+      let log = try
+        Hashtbl.find log_files path
+      with Not_found ->
+        let empty = Buffer.create 16 in
+        Hashtbl.add log_files path empty;
+        empty
+      in
+      Buffer.add_string log txt;
+      []
+    | _ -> raise Not_implemented
+
+  let flush_log path txt =
+    let dir = Filename.dirname path in
+    let base = Filename.basename path in
+    let pid = string_of_int (Unix.getpid ()) in
+    let new_path = Filename.concat dir (pid ^ "_" ^ base) in
+    let flags = [Open_wronly; Open_append; Open_creat; Open_text] in
+    let out_chan = open_out_gen flags 0o666 new_path in
+    let fd = Unix.descr_of_out_channel out_chan in
+    Unix.lockf fd Unix.F_LOCK 0;
+    output_string out_chan (Buffer.contents txt);
+    close_out out_chan
+
+  let flush_logs () =
+    Hashtbl.iter flush_log log_files
+
+  let hook_stat _ _ _ _ _ = raise Not_implemented
+  let hook_lstat _ _ _ _ _ = raise Not_implemented
+  let hook_opendir _ _ _ _ _ = raise Not_implemented
+  let hook_system c _ _ _ _ = match c with
+    | [String path] ->
+      let ic, oc = Unix.open_process path in
+      let buf_out = Buffer.create 4096
+      and buf_err = Buffer.create 128 in
+      (try
```

Comment:
not really sure what's going on here but this looks incredibly suspicious and is probably not the right way to read all the output of a process. Cf. https://stackoverflow.com/questions/29503960/in-ocaml-how-to-get-stdout-string-from-subprocess for a better solution and also bear in mind you need to use open_process_full so you get access to stderr as well.

Reply:
I understand better what is going on here but can you increase the number of characters read at one time to 4096?